File: src/main/java/ru/ezhov/test/tree/task/Application.java
package ru.ezhov.test.tree.task;

import org.codehaus.jackson.map.ObjectMapper;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Application {
    public static void main(String[] args) throws IOException {
        List<BinaryObject> binaryObjects = Arrays.asList(
                new BinaryObject(103, "Дополнительная выкладка", null),
                new BinaryObject(100, "Лояльность Хаммер", 103),
                new BinaryObject(101, "Хаммер", 103),
                new BinaryObject(111, "Хаммер child", 101),
                new BinaryObject(123, "ЮХУ", null)
        );

        Map<Integer, TreePresenter> treePresenterMap = new HashMap<>();

        List<TreePresenter> parentPresenters = new ArrayList<>();

        for (BinaryObject binaryObject : binaryObjects) {
            final Integer prnId = binaryObject.getPrnId();
            if (prnId == null) {
                final TreePresenter treePresenterParent = new TreePresenter(binaryObject.getId(), binaryObject.getName());
                treePresenterMap.put(binaryObject.getId(), treePresenterParent);
                parentPresenters.add(treePresenterParent);
            } else {
                final TreePresenter treePresenterParent = treePresenterMap.get(prnId);
                if (treePresenterParent != null) {
                    final TreePresenter presenter = new TreePresenter(binaryObject.getId(), binaryObject.getName());
                    treePresenterParent.add(presenter);
                    treePresenterMap.put(binaryObject.getId(), presenter);
                } else {
                    TreePresenter treePresenterParentNew = new TreePresenter(binaryObject.getId(), binaryObject.getName());
                    treePresenterMap.put(binaryObject.getId(), treePresenterParentNew);
                    parentPresenters.add(treePresenterParentNew);
                }
            }
        }

        ObjectMapper objectMapper = new ObjectMapper();
        System.out.println(objectMapper.writerWithDefaultPrettyPrinter().writeValueAsString(parentPresenters));
    }
}

class BinaryObject {
    private final int id;
    private final String name;
    private final Integer prnId;

    public BinaryObject(int id, String name, Integer prnId) {
        this.id = id;
        this.name = name;
        this.prnId = prnId;
    }

    public int getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public Integer getPrnId() {
        return prnId;
    }

    @Override
    public String toString() {
        return "BinaryObject{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", prnId=" + prnId +
                '}';
    }
}

class TreePresenter {
    private final int id;
    private final String text;
    private boolean checked = false;
    private boolean expanded = false;
    private boolean leaf = true;

    private List<TreePresenter> data = new ArrayList<>();

    public TreePresenter(int id, String text) {
        this.id = id;
        this.text = text;
    }

    public int getId() {
        return id;
    }

    public String getText() {
        return text;
    }

    public boolean isChecked() {
        return checked;
    }

    public boolean isExpanded() {
        return expanded;
    }

    public boolean isLeaf() {
        return leaf;
    }

    public void add(TreePresenter treePresenter) {
        data.add(treePresenter);
        this.leaf = data.isEmpty();
    }

    public List<TreePresenter> getData() {
        return data;
    }

    @Override
    public String toString() {
        return "TreePresenter{" +
                "id=" + id +
                ", text='" + text + '\'' +
                ", checked=" + checked +
                ", expanded=" + expanded +
                ", leaf=" + leaf +
                ", data=" + data +
                '}';
    }
}

File: src/main/kotlin/ru/ezhov/test/tree/task/Application.kt
package ru.ezhov.test.tree.task;

import org.codehaus.jackson.map.ObjectMapper
import java.util.*
import kotlin.collections.HashMap

fun main(args: Array<String>) {
    val binaryObjects: List<BinaryObjectK> = listOf(
            BinaryObjectK(103, "Дополнительная выкладка", null),
            BinaryObjectK(100, "Лояльность Хаммер", 103),
            BinaryObjectK(101, "Хаммер", 103),
            BinaryObjectK(111, "Хаммер child", 101),
            BinaryObjectK(123, "ЮХУ", null)
    )

    val treePresenterMap: MutableMap<Int, TreePresenterK> = HashMap()
    val parentPresenters: MutableList<TreePresenterK> = ArrayList()
    for (binaryObject in binaryObjects) {
        val prnId: Int? = binaryObject.prnId
        if (prnId == null) {
            val treePresenterParent = TreePresenterK(binaryObject.id, binaryObject.name)
            treePresenterMap[binaryObject.id] = treePresenterParent
            parentPresenters.add(treePresenterParent)
        } else {
            val treePresenterParent = treePresenterMap[prnId]
            if (treePresenterParent != null) {
                val presenter = TreePresenterK(binaryObject.id, binaryObject.name)
                treePresenterParent.add(presenter)
                treePresenterMap[binaryObject.id] = presenter
            } else {
                val treePresenterParentNew = TreePresenterK(binaryObject.id, binaryObject.name)
                treePresenterMap[binaryObject.id] = treePresenterParentNew
                parentPresenters.add(treePresenterParentNew)
            }
        }
    }

    val objectMapper = ObjectMapper()
    println(objectMapper.writerWithDefaultPrettyPrinter().writeValueAsString(parentPresenters))
}

class BinaryObjectK(val id: Int, val name: String, val prnId: Int?)

class TreePresenterK(
        val id: Int,
        val text: String,
        var checked: Boolean = false,
        var expanded: Boolean = false,
        var leaf: Boolean = true
) {
    val data: MutableList<TreePresenterK> = ArrayList()

    fun add(child: TreePresenterK) {
        data.add(child)
        this.leaf = data.isEmpty()
    }
}